interfaces/builtin: add gpio-keys interface for accessing events

bergotorino opened the pull request:
This new interface makes possible to read events coming from GPIO keys
such as some sort of buttons available on devices. The purpose is to
allow events to be captured by snaps that can trigger actions based on
them.

Thanks for helping us make a better snapd!
Have you signed the [license agreement](https://www.ubuntu.com/legal/contributors) and read the [contribution guide](https://github.com/snapcore/snapd/blob/master/CONTRIBUTING.md)?

zyga commented on interfaces/builtin/gpio_keys.go:
```diff
@@ -0,0 +1,89 @@
+	"github.com/snapcore/snapd/interfaces/udev"
+)
+
+const gpioKeysSummary = `allows access to gpio keys events`
```
Nitpick: use singular `keys` since the `events` are already plural.

zyga commented on interfaces/builtin/gpio_keys.go:
```diff
@@ -0,0 +1,89 @@
+
+const gpioKeysSummary = `allows access to gpio keys events`
+
+const gpioKeysBaseDeclarationSlots = `
```
The policy looks fine, you will need to add a new plug to a spread test that exercises all interface types (spread will fail in this so you will know what to do).

zyga commented on interfaces/builtin/gpio_keys.go:
```diff
@@ -0,0 +1,89 @@
+const gpioKeysConnectedPlugAppArmor = `
+# Description: Allow reading and writing events to gpio keys devices
+
+#
```
Perhaps discard the `evdev-based interface` fragment since the description below is more than adequate.

zyga commented on interfaces/builtin/gpio_keys.go:
```diff
@@ -0,0 +1,89 @@
+// to ensure that the device cgroup is in effect even when there are no
+// gpio keys present so that we don't give away all input to the snap.
+var gpioKeysConnectedPlugUDev = []string{
+	`KERNEL=="event[0-9]*", SUBSYSTEM=="input", ENV{ID_PATH}=="platform-gpio-keys"`,
```
This looks ok but please evaluate if we should use ID_PATH or ID_PATH_TAG, I don't know what the difference is from the top of my head.

zyga commented on interfaces/builtin/gpio_keys.go:
```diff
@@ -0,0 +1,89 @@
+}
+
+func (iface *gpioKeysInterface) UDevConnectedPlug(spec *udev.Specification, plug *interfaces.ConnectedPlug, slot *interfaces.ConnectedSlot) error {
+	spec.TriggerSubsystem("input")
```
This looks fine. CC @jdstrand

estutzenberger replied:
FWIW, `udevadm info --query=property --name=/dev/input/by-path/platform-gpio-keys-event` gives the following on the system:
```
DEVLINKS=/dev/input/by-path/platform-gpio-keys-event
DEVNAME=/dev/input/event1
DEVPATH=/devices/platform/gpio-keys/input/input1/event1
ID_INPUT=1
ID_INPUT_KEY=1
ID_PATH=platform-gpio-keys
ID_PATH_TAG=platform-gpio-keys
MAJOR=13
MINOR=65
SUBSYSTEM=input
USEC_INITIALIZED=11131064
net.ifnames=0
```

zyga commented on interfaces/builtin/gpio_keys.go:
```diff
@@ -0,0 +1,89 @@
+
+func init() {
+	registerIface(&gpioKeysInterface{commonInterface{
+		name:                  "gpio-keys",
```
The name is perhaps most interesting. Those are not GPIOs as exposed in sysfs. Perhaps to avoid confusing with the GPIO interface we could call this interface `platform-keys` or even `platform-input`. CC @niemeyer

bergotorino commented:
Updated the comments as asked. The tests will come next.

bergotorino commented:
Added tests to make sure interface works as expected and that keyboard devices cannot be accessed. This addresses points a) and b) (hopefully) from the @jdstrand's comment.

bergotorino commented:
Having addressed a few of the comments so far I think I need more guidance on how to complete all of them. @jdstrand could you review and provide suggestions? Thanks in advance!

bergotorino commented on interfaces/builtin/device_buttons.go:
```diff
@@ -0,0 +1,92 @@
+}
+
+func (iface *deviceButtonsInterface) UDevConnectedPlug(spec *udev.Specification, plug *interfaces.ConnectedPlug, slot *interfaces.ConnectedSlot) error {
+	spec.TriggerSubsystem("input")
```
Thanks, this is a good catch. I have already made the requested change.

bergotorino commented on tests/main/security-dev-input-event-denied/task.yaml:
```diff
@@ -63,3 +65,37 @@ execute: |
     # the device cgroup.
     MATCH "Operation not permitted" < call.error
+
+    # 2. Device Buttons
```
Good catch, done!

bergotorino commented on interfaces/builtin/device_buttons_test.go:
```diff
@@ -0,0 +1,106 @@
+	c.Assert(spec.Snippets(), testutil.Contains, `# device-buttons
+KERNEL=="event[0-9]*", SUBSYSTEM=="input", ENV{ID_INPUT_KEY}=="1", ENV{ID_INPUT_KEYBOARD}!="1", TAG+="snap_consumer_app"`)
+	c.Assert(spec.TriggeredSubsystems(), DeepEquals, []string{"input"})
+}
```
Done

zyga commented on interfaces/udev/udev.go:
```diff
@@ -64,6 +64,15 @@ func ReloadRules(subsystemTriggers []string) error {
+			// subsystem for input keys, then trigger the keys
+			// events in a way that is specific to input keys
+			// to not block other inputs.
+			output, err = exec.Command("udevadm", "trigger", "--property-match=ID_INPUT_KEY=1", "--property-match=ID_INPUT_KEYBOARD!=1").CombinedOutput()
```
This looks good, thanks!

zyga commented:
```

----------------------------------------------------------------------
FAIL: device_buttons_test.go:100: DeviceButtonsInterfaceSuite.TestStaticInfo

device_buttons_test.go:104:
    c.Assert(si.Summary, Equals, `allows access to device buttons as input event`)
... obtained string = "allows access to device buttons as input events"
... expected string = "allows access to device buttons

```

bboozzoo commented:
@jdstrand the PR was updated, can you take another look?

jdstrand commented on tests/main/security-dev-input-event-denied/task.yaml:
```diff
@@ -63,3 +73,30 @@ execute: |
+        exit 1
+    fi
+    # AppArmor is 'Permission denied' which is expected with default policy
+    MATCH "Operation not permitted" < call.error
```
The comment here doesn't go with the MATCH statement and what this is showing is that you have a device cgroup denial. You should disconnect the joystick interface (with a comment) before ' # 2. Device Buttons', then use:
```
MATCH "Permission denied" < call.error
```

bboozzoo replied:
I fixed that and ran into issues. With `device-buttons` connected, none of the devices matched, hence `snap-confine` did not put itself under device cgroup in the first place. As a result, the test snap had unconfined access to all input devices. I'll address that in a separate PR.

jdstrand replied:
This was why I wanted the spread tests! :) Thanks for finding this. In this PR, please add "/dev/full" to the cgroup like we did in the joystick interface here: https://github.com/snapcore/snapd/pull/5231/files

If/when the unconditional cgroup is added (ie, your other PR), then this /dev/full here and in the joystick interface can be removed.

bboozzoo commented:
@jdstrand pushed the workaround you suggested. Please take another look.

jdstrand commented on interfaces/builtin/device_buttons.go:
```diff
@@ -0,0 +1,93 @@
+}
+
+func (iface *deviceButtonsInterface) UDevConnectedPlug(spec *udev.Specification, plug *interfaces.ConnectedPlug, slot *interfaces.ConnectedSlot) error {
+	spec.TriggerSubsystem("key")
```
This should be "input/key" since that is what the change to udev.go is looking for. Sorry for not noticing this sooner.

jdstrand commented on tests/main/security-dev-input-event-denied/task.yaml:
```diff
@@ -63,3 +73,36 @@ execute: |
+    udevadm settle
+
+    # Note, '-event-kbd' devices aren't device buttons (those are
+    # -gpio-keys-even (evdev event*) and therefore shouldn't be added to
```
Typo: -gpio-keys-even**t**
